#PRÁCTICA UD 1 PIA - Roberto Martínez Soto#

Para la realización de esta práctica he seleccionado el dataset _diamonds_, que contiene datos sobre decenas de miles de diamantes con sus características y precios.
Considero que este dataset es lo suficientemente grande como para llevar a cabo el análisis de sus datos, así como para extraer información interesante.








###Primera inspección de los datos
Se ejecutado el siguiente código para visualizar los 10 primeros valores del dataset y posteriormente ver los tipos de datos del dataframe que he utilizado para cargar los datos.

In [12]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
print(sns.get_dataset_names())
df = sns.load_dataset("diamonds")
df.head(10)

['anagrams', 'anscombe', 'attention', 'brain_networks', 'car_crashes', 'diamonds', 'dots', 'dowjones', 'exercise', 'flights', 'fmri', 'geyser', 'glue', 'healthexp', 'iris', 'mpg', 'penguins', 'planets', 'seaice', 'taxis', 'tips', 'titanic']


carat        cut color clarity  depth  table  price     x     y     z
0   0.23      Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21    Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23       Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29    Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31       Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75
5   0.24  Very Good     J    VVS2   62.8   57.0    336  3.94  3.96  2.48
6   0.24  Very Good     I    VVS1   62.3   57.0    336  3.95  3.98  2.47
7   0.26  Very Good     H     SI1   61.9   55.0    337  4.07  4.11  2.53
8   0.22       Fair     E     VS2   65.1   61.0    337  3.87  3.78  2.49
9   0.23  Very Good     H     VS1   59.4   61.0    338  4.00  4.05  2.39

In [8]:
df.info()
print("Total de datos en el dataframe: " + str(len(df)))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53940 entries, 0 to 53939
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   carat    53940 non-null  float64 
 1   cut      53940 non-null  category
 2   color    53940 non-null  category
 3   clarity  53940 non-null  category
 4   depth    53940 non-null  float64 
 5   table    53940 non-null  float64 
 6   price    53940 non-null  int64   
 7   x        53940 non-null  float64 
 8   y        53940 non-null  float64 
 9   z        53940 non-null  float64 
dtypes: category(3), float64(6), int64(1)
memory usage: 3.0 MB
Total de datos en el dataframe: 53940


#1 - EXPLICACIÓN Y CONTEXTUALIZACIÓN DE LOS DATOS
Campos del dataset _diamonds_ y su significado:


*   ```carat``` : Peso del diamante en gramos.
*   ```cut``` : Calidad del corte, distinguiendo entre cinco categorías:
    * Fair
    * Good
    * Very Good
    * Premium
    * Ideal
*   ```color``` : Calificación del color considerando el nivel de transparencia.  
*   ```clarity``` : Representa la claridad del diamante, que es mayor cuantas menos imperfecciones tiene. Este campo abarca desde la mejor claridad, representada con el valor IF(Internally Flawless - sin imperfecciones internas, hasta el peor nivel de claridad, que es representado por el valor I3(imperfecciones visibles a simple vista).
*   ```table``` : Indica la calidad del corte, representando una correlación entre el ancho de la parte superior del diamante y su anchura máxima, expresado en un porcentaje. Este valor es relevante en cuanto a que afecta a cómo se va a comportar la luz una vez incide a través de la parte superior de la gema y se refleja internamente.

   ![Tabla](https://www.brilliance.com/front/img/table_illustration.1d5c5a3f.jpeg)
*   ```price``` : Precio en dólares estadounidenses.
*   ```x``` : Longitud en milímetros.
*   ```y``` : Anchura en milímetros.
*   ```z``` : Profundidad en milímetros.
*   ```depth``` : Expresa una relación porcentual entre el ancho del diamante y la altura, medida desde la mesa hasta el vértice inferior.

   ![Depth](https://www.brilliance.com/front/img/depth_illustration.3cfc28e7.jpeg)


 Tras la primera inspección del dataframe se puede observar que todos sus campos contienen datos y no va a haber presencia de nulos, lo que favorece en gran medida el análisis de los mismos.

 Para enriquecer el análisis inicial se realiza un primer vistazo a nivel estadístico del dataframe utilizando la función _describe_ que ofrece la librería _seaborn_, así como calculando el coeficiente de de variación de cada columna.









In [3]:
df.describe()

carat         depth         table         price             x  \
count  53940.000000  53940.000000  53940.000000  53940.000000  53940.000000   
mean       0.797940     61.749405     57.457184   3932.799722      5.731157   
std        0.474011      1.432621      2.234491   3989.439738      1.121761   
min        0.200000     43.000000     43.000000    326.000000      0.000000   
25%        0.400000     61.000000     56.000000    950.000000      4.710000   
50%        0.700000     61.800000     57.000000   2401.000000      5.700000   
75%        1.040000     62.500000     59.000000   5324.250000      6.540000   
max        5.010000     79.000000     95.000000  18823.000000     10.740000   

                  y             z  
count  53940.000000  53940.000000  
mean       5.734526      3.538734  
std        1.142135      0.705699  
min        0.000000      0.000000  
25%        4.720000      2.910000  
50%        5.710000      3.530000  
75%        6.540000      4.040000  
max       58.900000     31.800000

In [4]:
df_estadisticas = df.describe()

media = df_estadisticas.loc['mean']
desviacion_estandar = df_estadisticas.loc['std']

porcentaje_desviacion = ((desviacion_estandar / media) * 100).round(2)
porcentaje_desviacion = porcentaje_desviacion.astype(str) + ' %'
print(porcentaje_desviacion)

carat      59.4 %
depth      2.32 %
table      3.89 %
price    101.44 %
x         19.57 %
y         19.92 %
z         19.94 %
dtype: object


#2 - ANÁLISIS DE LAS COLUMNAS


Teniendo en cuenta la información obtenida en el primer punto, se pueden tomar en consideración ciertas cuestiones:

Tanto la columna ```carat``` como la columna ```price``` presentan una gran variación entre su media y el resto de valores, alcanzando hasta el 101,44% en el caso del precio. Este dato implica que para estos dos campos, la media de sus valores no es tan significativa, al contrario de lo que sucede con ```depth``` y ```table```.
Si nos fijamos en el coeficiente de variación de ```carat``` y ```price```, tratándose de un _dataset_ con tan pocas columnas, se puede deducir que más allá de lo que dicta la lógica, estos valores estarán relacionados. Cuanto mayor sea el peso de un diamante, mayor será su precio de mercado.

Con este primer análisis se puede deducir que la información más interesante que puede ofrecer este _dataset_ es la relación de distintos parámetros con respecto al precio de los diamantes.

Para continuar, se ha de tener en cuenta la relevancia de las distintas propiedades que contiene el _dataset_.
Cinco de ellas corresponden a medidas de los diamantes, y entre esas cinco se podrían considerar más importantes aquellas que tienen mayor relación con el precio. Sabiendo que tanto ```depth``` como ```table``` ofrecen información sobre las proporciones y características lumínicas de las gemas y teniendo en cuenta que dichas propiedades son cálculos obtenidos de las medidas de las mismas, los datos de las columnas ```x```, ```y``` y ```z``` no ofrecen información relevante, con lo que se procede a eliminarlos.



In [14]:
# prompt: Quiero eliminar las columnas "x" "y" y "z" del dataframe df

# Eliminar las columnas 'x', 'y' y 'z' del DataFrame df
df = df.drop(['x', 'y', 'z'], axis=1)
df

carat        cut color clarity  depth  table  price
0       0.23      Ideal     E     SI2   61.5   55.0    326
1       0.21    Premium     E     SI1   59.8   61.0    326
2       0.23       Good     E     VS1   56.9   65.0    327
3       0.29    Premium     I     VS2   62.4   58.0    334
4       0.31       Good     J     SI2   63.3   58.0    335
...      ...        ...   ...     ...    ...    ...    ...
53935   0.72      Ideal     D     SI1   60.8   57.0   2757
53936   0.72       Good     D     SI1   63.1   55.0   2757
53937   0.70  Very Good     D     SI1   62.8   60.0   2757
53938   0.86    Premium     H     SI2   61.0   58.0   2757
53939   0.75      Ideal     D     SI2   62.2   55.0   2757

[53940 rows x 7 columns]

Con respecto a las columnhas con datos categóricos, que son ```cut``` y ```color``` se ha decidido respetar sus datos para realizar un análisis correlativo con el resto de propiedades del diamante y observar si pueden ser relevantes y ofrecer algún tipo de información.

Hasta el momento se han analizado aspectos al respecto de las diferentes columnas existentes, pero también merece la pena prestarle atención a las filas, pues al tratarse de un gran volumen de datos, puede haber duplicados.
Se puede comprobar con la función `duplicated()` de la librería _pandas_.

In [16]:
duplicados = df.duplicated()
num_duplicados = duplicados.sum()
print(f"Número de filas duplicadas: {num_duplicados}")

Número de filas duplicadas: 803


Tras ejecutar el código se observa que hay presentes 803 filas duplicadas.
Es en este punto donde cabe plantearse ciertas preguntas:
¿Cómo llegan los datos de estos diamantes al dataset?¿Se dan los datos de alta en un excel directamente por parte de un trabajador o investigador?
En el caso de que la carga de los datos se realice desde un programa informático, ¿Contempla dicho programa la posibilidad de detectar duplicidades antes de incluir los datos?

Se pueden plantear una gran cantidad de preguntas al respecto para conocer la naturaleza de los datos y con ello tomar la decisión de conservar todos los datos o eliminar los duplicados.

Como ejemplo de caso, si está trabajando en un contrato realizado con una empresa, seguramente se podrá plantear alguna de estas cuestiones al cliente, y una vez se cuente con alguna información, se pueden realizar ajustes en los datos que nos faciliten realizar una mejora en el análisis de los mismos y la posterior creación de un mejor modelo de IA. O en el caso de que no se cuente con suficiente información, se puede señalar al cliente el 'punto de dolor' que enfrenta en términos de sus propios datos, con el fin de definir mejor el alcance de la solución informática que se debe desarrollar."

Dado que en este caso no se cuenta con mucha información respecto a las cuestiones planteadas y considerando que los diamantes no tienen un identificador, no es posible deducir que cada fila corresponda a un diamante único. Además, los duplicados representan solo un 1,48% del total, con lo que se desestima la posibilidad de eliminarlos.

En el caso de que se tomara la decisión de eliminarlos, se podría realizar con la función `drop_duplicates()` de la librería _pandas_.